### DISCUSSION TITLE
Stop hiding function pointer delegate constructor

### DISCUSSION THREAD

**Discussion by ThadHouse** [dotnet/csharplang#4871] [Language Ideas]
Since we now have full support for Function Pointers in C#, we should stop hiding the function pointer taking delegate constructor that exists on all delegates. This would allow some interesting delegate to function pointer mapping scenarios.

**Comment by bernd5** [comment]
That would be really great!

I wrote this as a workaround:
```csharp
 public static DelegateType CreateDelegate<DelegateType>(object target, IntPtr methodPtr)
      where DelegateType : Delegate
{
    var ctor = typeof(DelegateType).GetConstructor(new[] { typeof(object), typeof(IntPtr) })!;
    var result = ctor.Invoke(new[] { target, methodPtr });
    return (DelegateType)result;
}
```

**Comment by tannergooding** [comment]
For reference, every delegate type inherits from `System.Delegate` or `System.MulticastDelegate` (the latter for those defined in C#).

These types expose the following public methods:
* `public MyDelegate(object @object, IntPtr method)`
* `public virtual void Invoke(<args>)`
* `public virtual System.IAsyncResult BeginInvoke(<args>, System.AsyncCallback callback, object @object);`
* `public virtual void EndInvoke(System.IAsyncResult result)`

The first is a constructor that takes the `Target` property and a function pointer (currently exposed as an `IntPtr`). Given function pointers now exist in C#, it might be beneficial to no longer hide this constructor and to simply allow `new MyDelegate(null, someFnptr)`.

**Comment by 333fred** [comment]
If anyone would like to take the time to propose the rule changes necessary to https://github.com/dotnet/csharplang/blob/main/spec/expressions.md#delegate-creation-expressions, I'd be willing to champion the feature.

**Comment by sakno** [comment]
Possibly dup of #3680
